runtime/gc_glue_java/ArrayletObjectModel.hpp

```diff
@@ -255,6 +267,66 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
 		return getSpineSizeWithoutHeader(layout, numberArraylets, dataSize, alignData);
 	}
 
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+	/**
+	 * Checks if arraylet falls into corner case of discontigous data
+	 * Arraylet possible cases:
+	 * 0: Empty arraylets, in this case the array is represented as
+	 *		an arraylet however it does not contain any data, but it
+	 *		does contain an arrayoid (leaf pointer) that points to NULL.
+	 *		Even though this case is represented as a discontiguous arraylet
+	 *		internally due to its implementation, it is actually a contiguous
+	 *		array with length zero.
+	 * 1: The total data size in arraylet is between 0 and region
+	 *		size. Small enough to make the arraylet layout contiguous,
+	 *		in which case this function is unreachable.
+	 * 2: The total data size in arraylet is exacly the same size
+	 *		of a region, in which case the arraylet will contain an
+	 *		extra NULL arrayoid. In this case we do not need to double
+	 *		map since we already have a contiguous representation of the
+	 *		data at first leaf.
+	 * 3: Similar to first case, the data portion is slightly smaller than
+	 *		a region size, however not small enough to include header and data
+	 *		at the same region to make it contiguous. In which case we would
+	 *		have one leaf, where we also do not need to double map.
+	 * 4: The total data size in arraylet is stricly greater than one region;
+	 *		however, not multiple of region size. Since with enabled double map
+	 *		layout is always discontiguous, we would have 2 or more arraylet leaves
+	 *		therefore we always double map.
+	 * 5: The total data size in arraylet is stricly greater than one region and
+	 *		multiple of region size. Here we would have 2 or more arraylet leaves
+	 *		containing data and the last leaf pointing to NULL. Nonetheless, we
+	 *		always double map in this case
+	 *
+	 * @param spine Pointer to an array indexable object spine
+	 * @return false in case corner cases 0, 2 or 3 are valid. On the other hand,
+	 *		if cases 4 or 5 are true, the function returns true.
+	 */
+	MMINLINE bool
+	isArrayletDataDiscontiguous(J9IndexableObject *spine)
+	{
+		UDATA arrayletLeafCount = numArraylets(spine);
+		UDATA sizeInElements = getSizeInElements(spine);
+		return ((arrayletLeafCount > 1) && !isOneArrayletLeafWithNULL(spine, arrayletLeafCount, sizeInElements));
```

Comment:
Please remove outermost parentheses (`return` is not a function - here and below).